xonsh/tools.py

```diff
@@ -1648,34 +1650,36 @@ def lazyget(self, key, default=None):
         return self._cmds_cache.get(key, default)
 
     def locate_binary(self, name):
-        """Locates an executable on the file system."""
+        """Locates an executable on the file system using the cache."""
+        possibilities = self.get_possible_names(name)
 
         if ON_WINDOWS:
+            from xonsh.dirstack import _get_cwd  # circular dependency
             # Windows users expect to be able to execute files in the same
             # directory without `./`
             cwd = _get_cwd()
 
-            local_bins = [full_name for full_name in self.get_possible_names(name)
-                          if os.path.isfile(full_name)]
-            if local_bins:
-                return os.path.abspath(os.path.relpath(local_bins[0], cwd))
+            local_bin = next((
+                full_name for full_name in possibilities
+                if os.path.isfile(full_name)
+            ), None)
 
-        possibilities = self.get_possible_names(name)
-        matches = possibilities & self.all_commands.keys()
+            if local_bin:
+                return os.path.abspath(os.path.relpath(local_bin, cwd))
```

Comment:
Should it not be enough to call `os.path.abspath(local_bin)`. What does the `relpath` do?